beacon-chain/rpc/beaconv1/blocks.go

```diff
@@ -24,20 +116,217 @@ func (bs *Server) ListBlockHeaders(ctx context.Context, req *ethpb.BlockHeadersR
+	}
+	if blk == nil {
+		return nil, status.Errorf(codes.NotFound, "Could not find requested block")
+	}
+
+	v1Block, err := migration.V1Alpha1ToV1Block(blk)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Could not convert block to v1")
+	}
+
+	return &ethpb.BlockResponse{
+		Data: &ethpb.BeaconBlockContainer{
+			Message:   v1Block.Block,
+			Signature: blk.Signature,
+		},
+	}, nil
 }
 
 // GetBlockRoot retrieves hashTreeRoot of BeaconBlock/BeaconBlockHeader.
 func (bs *Server) GetBlockRoot(ctx context.Context, req *ethpb.BlockRequest) (*ethpb.BlockRootResponse, error) {
-	return nil, errors.New("unimplemented")
+	var root []byte
+	var err error
+	switch string(req.BlockId) {
+	case "head":
+		root, err = bs.HeadFetcher.HeadRoot(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve head block: %v", err)
+		}
+		if root == nil {
+			return nil, status.Errorf(codes.NotFound, "No head root was found")
+		}
+	case "finalized":
+		finalized := bs.FinalizationFetcher.FinalizedCheckpt()
+		if finalized == nil {
+			return nil, status.Errorf(codes.NotFound, "No finalized root was found")
+		}
+		root = finalized.Root
+	case "genesis":
+		blk, err := bs.BeaconDB.GenesisBlock(ctx)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for genesis slot: %v", err)
+		}
+		if blk == nil {
+			return nil, status.Error(codes.NotFound, "Could not find genesis block")
+		}
+		blkRoot, err := blk.Block.HashTreeRoot()
+		if err != nil {
+			return nil, status.Error(codes.Internal, "Could not hash genesis block")
+		}
+		root = blkRoot[:]
+	default:
+		if len(req.BlockId) == 32 {
+			block, err := bs.BeaconDB.Block(ctx, bytesutil.ToBytes32(req.BlockId))
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not retrieve block for block root %#x: %v", req.BlockId, err)
+			}
+			if block == nil {
+				return nil, status.Error(codes.NotFound, "Could not find any blocks with given root")
+			}
+
+			root = req.BlockId
+		} else {
+			slotInt, err := strconv.Atoi(string(req.BlockId))
+			if err != nil {
+				return nil, errors.Wrap(err, "could not decode block id")
+			}
+			slot := uint64(slotInt)
+			roots, err := bs.BeaconDB.BlockRoots(ctx, filters.NewFilter().SetStartSlot(slot).SetEndSlot(slot))
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "Could not retrieve blocks for slot %d: %v", slot, err)
+			}
+
+			numRoots := len(roots)
+			if numRoots == 0 {
+				return nil, status.Error(codes.NotFound, "Could not find any blocks with given root")
+			}
+			root = roots[0][:]
+			for _, blockRoot := range roots {
+				canonical, err := bs.CanonicalFetcher.IsCanonical(ctx, blockRoot)
+				if err != nil {
+					return nil, status.Errorf(codes.Internal, "Could not determine if block root is canonical: %v", err)
+				}
+				if canonical {
+					root = blockRoot[:]
```

Comment:
```suggestion
					root = blockRoot[:]
					break
```

No need to keep checking in this list once you have found it.

Reply:
Fixed